### DISCUSSION TITLE
Add section in config file to specify command line options

### DISCUSSION THREAD

**Discussion by FStefanni** [clangd/clangd#1782] [Ideas]
Hi,

this follows the feedback of issue https://github.com/clangd/clangd/issues/1781#issuecomment-1756945039.
The idea is simple: in the config file, it would be nice to have a way to specify options currently possible via command line, but not somehow present in the config file.
For example, a dedicated section could be added, something as "ExtraCliOptions" or "ExtraToolOptions".

Why this is useful?
Because in general this increases the flexibility of how people uses the tool.
For example, in my case it is not so convenient/easy to configure the tool via command line for each different project (and each project requires different options). On the other way, just copy the config file inside the project, and modify it is just straightforward.

My general view (not only regarding clangd, but any tool), is that the best would be to always to be able to pass *all* the options
in any of the possible ways (e.g. config file, cli, env vars, etc.).
This is so convenient, for example:

* to run something "spot" a command line option is the fastest way
* to run a "session" on the cli, an env var is nice
* to share the configuration with other people or "store the configuration", the config file is the best

This is sometimes not possible (or too much), for example some complex options are hard to  be described by using the command line or env var, so maybe they are only possible via config file.

In the specific case of clangd, it would be very useful (and probably more than enough) to be able to specify whatever command line option also via config file. 

Regards

**Comment by HighCommander4** [comment]
One thing to note here is that command line arguments apply to every file opened by that clangd process, while config file entries are scoped to the directory containing the config file (for [project config](https://clangd.llvm.org/config.html#files) files), and can additionally be scoped to a subset of that directory using [`If`](https://clangd.llvm.org/config.html#if).

So, we would have to make a choice here:

 1. Either we allow scoping all command-line options to a directory as well (which means the place in the code where we check the flag has to change from just checking a global flag, to checking whether the file we're operating on is in the scope of the option).
     * Some command-line options, like `--malloc-trim`, are not specific to operations on a file. It's unclear what scoping this to a directory would even mean.
 2. Or, we invent a new category of config file entry which is not scoped to a directory. Which I guess means it can only go in the [user config](https://clangd.llvm.org/config.html#files) file, not a project config file, since project config files are implicitly scoped to the project directory? And additionally it cannot appear in an `If` block? (What happens if you try putting it in an `If` block?)

**Comment by FStefanni** [comment]
Hi,

I vote for option 1! It seems the simplest and more natural.
I mean: 
* Scoped as usual for almost everything
* Special options, as `--malloc-trim`, simply will be kicked in if selected and in that case they will work globally, as expected.

Regards

**Comment by HighCommander4** [comment]
A possible alternative formulation to `ExtraToolOptions`:

 1. For command-line options which it makes sense to scope to a directory, add an equivalent config option. For example, for `--header-insertion=never`, add `HeaderInsertion: Never`.
 2. Add a command-line flag `--config-fragment=<fragment>`, which interprets `<fragment>` as a fragment of the user config file. The flag can be specified multiple times, with the effect of adding multiple user-config fragments.
 3. The original command-line option, e.g. `--header-insertion=never`, becomes an alias for e.g. `--config-fragment="HeaderInsertion: Never"`. (Perhaps eventually the original form can be deprecated and removed.)
 4. Options which it does not make sense to scope to a directory, like `--malloc-trim`, are left alone.

**Reply by HighCommander4** [reply]
>    1. For command-line options which it makes sense to scope to a directory, add an equivalent config option. For example, for `--header-insertion=never`, add `HeaderInsertion: Never`.

This has been done for a couple of options:

 * `--header-insertion` (https://github.com/clangd/clangd/issues/2032)
 * `--function-arg-placeholders` (https://github.com/llvm/llvm-project/issues/63565)

Where it makes sense, PRs porting over other options are welcome.

>    2. Add a command-line flag `--config-fragment=<fragment>`, which interprets `<fragment>` as a fragment of the user config file. The flag can be specified multiple times, with the effect of adding multiple user-config fragments.

This is tracked in https://github.com/clangd/clangd/issues/2493.

**Comment by FStefanni** [comment]
This seems also fine.
I still not agree to point 4: it seems to make sense to enable this behavior depending on the project we are going to open in the IDE, so again it could be in a config file.
